evalml/automl/engine/engine_base.py

```diff
@@ -0,0 +1,135 @@
+import sys
+import time
+import traceback
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+
+import numpy as np
+import pandas as pd
+
+from evalml.automl.utils import tune_binary_threshold
+from evalml.exceptions import PipelineScoreError
+from evalml.model_family import ModelFamily
+from evalml.preprocessing import split_data
+from evalml.problem_types import is_binary, is_multiclass
+from evalml.utils.logger import get_logger
+from evalml.utils.woodwork_utils import _convert_woodwork_types_wrapper
+
+logger = get_logger(__file__)
+
+
+class EngineBase(ABC):
+    """Base class for the engine API which handles the fitting and evaluation of pipelines during AutoML."""
+
+    def __init__(self, X_train=None, y_train=None, automl=None, should_continue_callback=None, pre_evaluation_callback=None, post_evaluation_callback=None):
+        """Base class for the engine API which handles the fitting and evaluation of pipelines during AutoML.
+
+        Arguments:
+            X_train (ww.DataTable): training features
+            y_train (ww.DataColumn): training target
+            automl (AutoMLSearch): a reference to the AutoML search. Used to access configuration and by the error callback.
+            should_continue_callback (function): returns True if another pipeline from the list should be evaluated, False otherwise.
+            pre_evaluation_callback (function): optional callback invoked before pipeline evaluation.
+            post_evaluation_callback (function): optional callback invoked after pipeline evaluation, with args pipeline and evaluation results. Expected to return a list of pipeline IDs corresponding to each pipeline evaluation.
+        """
+        self.X_train = X_train
+        self.y_train = y_train
+        self.automl = automl
+        self._should_continue_callback = should_continue_callback
+        self._pre_evaluation_callback = pre_evaluation_callback
+        self._post_evaluation_callback = post_evaluation_callback
+
+    @abstractmethod
+    def evaluate_batch(self, pipelines):
+        """Evaluate a batch of pipelines using the current dataset and AutoML state.
+
+        Arguments:
+            pipeline_batch (list(PipelineBase)): A batch of pipelines to be fitted and evaluated
+
+        Returns:
+            list (int): a list of the new pipeline IDs which were created by the AutoML search.
+        """
+
+    @staticmethod
+    def train_and_score_pipeline(pipeline, automl, full_X_train, full_y_train):
```

Comment:
We are passing `X_train` from the automl obj -- see `SequentialEngine.evaluate_batch`. The variable name here is not importance since this is a static method.

Reply:
Ah if you're asking why not use `automl.X_train` here, yeah, I would like us to eventually not have a direct reference to the automl search in this method. Why: because this code will eventually be running on a remote worker.

Reply:
Mmm, got it, that makes sense. Thanks for clarifying!